### PR TITLE
Adaround support

### PR DESCRIPTION
**Pull request by Giuseppe5**
Current functionalities:

- Learned hard sigmoid implementation
- Supports for all different temperature settings for the loss
- Caching of the input to speed-up layer-wise learning
- Combining layers and subsequent modules (i.e., batch norm, activations, etc.)

Missing:

- Handling of all the arguments (DataLoader, optimizer, adaround-specific params)
- Other implementations for the learnable parameter
- Other implementation of the loss

### CODE REVIEW

File: src/brevitas/core/function_wrapper/ops_ste.py [hunk 1]
```diff
@@ -38,6 +38,34 @@ def forward(self, x: torch.Tensor):
         return floor_ste(x)
 
 
+class AdaRoundSte(brevitas.jit.ScriptModule):
+    """
+    This Module implements AdaRound representation, where each weight has a learnable parameter
+    that decides if "ceil" or "floor" rounding type has to be used.
+    """
+
+    def __init__(self, shape: int) -> None:
```
Comment (Giuseppe5) [inline src/brevitas/core/function_wrapper/ops_ste.py:47]: change to tracked_parameter_list.
Check if tracked_parameter_list is 1, otherwise raise error.
tracked_parameter_list[0].shape

File: src/brevitas/core/function_wrapper/ops_ste.py [hunk 2]
```diff
@@ -38,6 +38,34 @@ def forward(self, x: torch.Tensor):
...
+    def hard_sigmoid(self) -> torch.Tensor:
+        """
+        During training, it produces soft-quantized results, bounded between 0 and 1.
+        During evaluation, it outputs 0 (indicating floor rounding) or 1 (indicating ceil rounding).
+        """
+        if self.training:
+            p = torch.sigmoid(self.value)
+            p = torch.clamp(p * (1.1 + 0.1) - 0.1, 0.0, 1.0)
+        else:
+            p = (self.value>0).float()
```
Comment (Giuseppe5) [inline src/brevitas/core/function_wrapper/ops_ste.py:66]: spacing self.value > 0

File: src/brevitas/core/function_wrapper/ops_ste.py [hunk 3]
```diff
@@ -38,6 +38,34 @@ def forward(self, x: torch.Tensor):
...
+    """
+
+    def __init__(self, shape: int) -> None:
+        super(AdaRoundSte, self).__init__()
+        self.value = torch.nn.Parameter(torch.full(shape, 0.))
+
+    @brevitas.jit.script_method
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        p = self.hard_sigmoid()
+        return floor_ste(x) + p
```
Comment (Giuseppe5) [inline src/brevitas/core/function_wrapper/ops_ste.py:54]: p.to(x.dtype)
